### PR TITLE
Adding expression function to retrieve anomalies 

### PR DESCRIPTION
**Pull request by amitgalitz**
### Description

Creates and registers an expression function to retrieve anomalies from AD as a vis-layer to be rendered on top of visualizations in OSD.

The expression function takes a detectorId argument and utilizes the date range in the context to fetch anomalies. The anomalies are then returned as vis-layer which visualize_embeddable in OSD renders onto the given visualization.

This PR is rebased directly from the featureAnywhere project. It doesn't have the changes to public/plugins.tsx as that code is yet unmerged. The only lines added there are: 

```
plugins.expressions.registerFunction(overlayAnomaliesFunction);
setClient(core.http);
```

### Check List
- [x] Commits are signed per the DCO using `--signoff`

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: opensearch_dashboards.json
Comment (ohltyler) [inline opensearch_dashboards.json:6]: nit / unrelated: while you're changing this file, I think we can go ahead and remove `navigation`. It was set when the plugin was initially created before this was understood and it isn't needed.

File: public/expressions/overlay_anomalies.ts
Comment (ohltyler) [inline public/expressions/overlay_anomalies.ts:10]: For new files we can use the short form:
```
/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */
```

File: public/expressions/overlay_anomalies.ts
Comment (ohltyler) [inline public/expressions/overlay_anomalies.ts:37]: nit: can be simplified to `'../utils/constants'`

File: public/services.ts
Comment (ohltyler) [inline public/services.ts:10]: Same as https://github.com/opensearch-project/anomaly-detection-dashboards-plugin/pull/448/files#r1153574259

**Comment by ohltyler** [comment]
This PR is rebased directly from the featureAnywhere project. It doesn't have the changes to public/plugins.tsx as that code is yet unmerged.

Any reason these changes can't be part of this PR? Otherwise after merging, the branch won't be able to use the expr fn since it isn't registered and `client` service isn't set.

File: opensearch_dashboards.json
Comment (ohltyler) [inline opensearch_dashboards.json:11]: this can be removed

File: public/expressions/overlay_anomalies.ts
Comment (ohltyler) [inline public/expressions/overlay_anomalies.ts:69]: I think we can remove this comment. I had added it initially when trying to develop it but it doesn't seem to add much value now

File: public/expressions/overlay_anomalies.ts
Comment (ohltyler) [inline public/expressions/overlay_anomalies.ts:92]: I wonder if we should extract this into a constant. It might make it easier to track when processing the values on OSD side.

File: public/expressions/overlay_anomalies.ts
Comment (ohltyler) [inline public/expressions/overlay_anomalies.ts:136]: A few notes on this part:
- `type` will be the header used when categorizing and grouping the `VisLayer`s on the view events flyout, so ideally it should be something like `Anomaly Detectors` (this could be pulled into a constant as well)
- `urlPath` should be the full path to the relevant details page for the plugin resource - in our case, `${PLUGIN_NAME}#/detectors/${detectorId}/results` probably

I'll follow up by adding some more descriptions on these fields in the interface definitions. It might be worth adding some comments wherever these are implemented here too to prevent confusion.

File: public/expressions/overlay_anomalies.ts
Comment (amitgalitz) [reply public/expressions/overlay_anomalies.ts:92]: Here also I wasn't 100% sure if it should be `anomaly-detection` or `anomaly-detection-dashboards` or even `anomalyDetectionDashboards` as we list on opensearch_dashboards.json. Definitely agree it should be a constant

File: public/expressions/overlay_anomalies.ts
Comment (ohltyler) [inline public/expressions/overlay_anomalies.ts:141]: nit: duplicate type declaration not needed since it's defined in `getAnomalies()`

File: public/expressions/overlay_anomalies.ts
Comment (ohltyler) [inline public/expressions/overlay_anomalies.ts:165]: can you add `TODO` to this comment so it can be more easily tracked?

File: public/expressions/overlay_anomalies.ts
Comment (ohltyler) [inline public/expressions/overlay_anomalies.ts:99]: Just for further documentation, could you add a comment for this function describing more about its purpose / how it's adding a vislayer onto an existing array of vislayers? Making that more clear can help us reference it when adding more documentation / tutorials on how other plugins may integrate.

File: public/services.ts
Comment (ohltyler) [inline public/services.ts:1]: nice! thanks for creating this file. Makes it cleaner for having a clear defined set of the dependencies being used from the required plugins.

File: public/expressions/overlay_anomalies.ts
Comment (ohltyler) [reply public/expressions/overlay_anomalies.ts:92]: The main purpose of this field is to provide another way to categorize the `VisLayer`s. Few advantages include:
1. Plugins having the opportunity to filter all `VisLayer`s that match the plugin's `originPlugin` field for a particular embeddable to post-process the data, instead of having to re-fetch saved objects (not sure if applicable given the current way plugins are consuming visualize_embeddable, but it leaves that option open)
2. More categorizing/filtering within the view events page or other OSD-owned components that are displaying the `VisLayer`s in some organized manner

Perhaps these `originPlugin` constants, and `type` constants can be put in a dedicated constants file specific to `VisLayer`s / this feature, so we can kind of split up constants that are only used within our plugin vs. constants that may be utilized in OSD via the render framework.

Comment (ohltyler) [reply public/expressions/overlay_anomalies.ts:92]: To clarify, this field isn't intended to be user-facing so it can be something internal like `anomalyDetectionDashboards` (this one makes sense to me to be consistent with the internal plugin name)

File: public/expressions/overlay_anomalies.ts
Comment (ohltyler) [reply public/expressions/overlay_anomalies.ts:136]: It's difficult to formalize this across plugins without limiting flexibility, since all we care about here is a string. I think we can leave this up to the plugins for now, as well as document the use of it in OSD. If that's still causing problems down the road we could revisit the interfaces.

File: public/plugin.tsx
Comment (ohltyler) [inline public/plugin.tsx:29]: looks like this is copied over from a separate PR, can these action-related code changes be omitted from this PR? They can be added to the file as part of that PR so the history is clean.

Comment (ohltyler) [reply public/plugin.tsx]: fixed in latest commit, thanks.

File: public/expressions/overlay_anomalies.ts
Comment (amitgalitz) [reply public/expressions/overlay_anomalies.ts:10]: will change

File: opensearch_dashboards.json
Comment (amitgalitz) [reply opensearch_dashboards.json:6]: removed, also added visAugmenter which I forgot to add originally here when simplifying the PR

File: public/expressions/overlay_anomalies.ts
Comment (amitgalitz) [reply public/expressions/overlay_anomalies.ts:81]: convertAnomaliesToPointInTimeEventsVisLayer makes sense to me as its the most accurate to what method is doing, good call.

File: public/expressions/overlay_anomalies.ts
Comment (amitgalitz) [reply public/expressions/overlay_anomalies.ts:136]: I am all for documenting as I understand the limitation around formalizing a specific standard. Good call adding some description in the interface definition. I will add some comment in this code and potentially we can also add more info in documentation for future plugin onboarding

File: public/expressions/overlay_anomalies.ts
Comment (amitgalitz) [reply public/expressions/overlay_anomalies.ts:92]: I guess though, do you want the originPlugin constants here for the `VisLayer` interface and the name for the `PluginResource` be exactly the same?

File: public/expressions/overlay_anomalies.ts
Comment (ohltyler) [reply public/expressions/overlay_anomalies.ts:136]: Forgot to also add:
`name` should be the name of the plugin resource, not the name of the plugin name. In this case, detector name

File: public/expressions/overlay_anomalies.ts
Comment (ohltyler) [reply public/expressions/overlay_anomalies.ts:92]: >  I guess this could be good to point out either in comments above the interface or to add to a plugin onboarding documentation of some sort

Yeah, that will be captured as part of https://github.com/opensearch-project/OpenSearch-Dashboards/issues/3383

File: public/expressions/overlay_anomalies.ts
Comment (amitgalitz) [reply public/expressions/overlay_anomalies.ts:120]: Added

File: public/expressions/overlay_anomalies.ts
Comment (amitgalitz) [reply public/expressions/overlay_anomalies.ts]: added this

File: public/expressions/overlay_anomalies.ts
Comment (ohltyler) [inline public/expressions/overlay_anomalies.ts:76]: Just to confirm, if there is failures coming from the cluster will this throw an error, or just store the error in `resp`? I Imagine we want to make sure an error is thrown so its caught in the `catch` block you have to process and produce an error `VisLayer`

Comment (amitgalitz) [reply public/expressions/overlay_anomalies.ts:76]: good catch, I actually fixed this by moving the call to getDetectorName inside the try, catch. Now if there is an error from the cluster it is automatically caught in the catch (like  status 4xx). If we get some response but no detector name, I manually throw the error which gets translated to a fetching failure.

Comment (ohltyler) [reply public/expressions/overlay_anomalies.ts:76]: It still seems a little hacky to be basing success/failure on if the name is empty or not. In fact, can we just use existing redux functions here to handle the errors for us? For example, [getDetector](https://github.com/opensearch-project/anomaly-detection-dashboards-plugin/blob/main/public/redux/reducers/ad.ts#L391)? We can then process any errors seen, similar to [this example](https://github.com/opensearch-project/anomaly-detection-dashboards-plugin/blob/main/public/pages/ReviewAndCreate/containers/ReviewAndCreate.tsx#L184-L199) with `then()`/`catch()` blocks. Lmk what you think.

Comment (ohltyler) [reply public/expressions/overlay_anomalies.ts:76]: Similar thought regarding the `anomalySummaryResponse` on L65. Maybe we can use the [searchResults](https://github.com/opensearch-project/anomaly-detection-dashboards-plugin/blob/main/public/redux/reducers/anomalyResults.ts#L124) redux fn.

Comment (ohltyler) [reply public/expressions/overlay_anomalies.ts:76]: I think the original intention of using `client()` standalone here was if other cluster calls needed to happen. But given that we only need standard ones we already have that are integrated with redux's already-consumed HTTP client, I think we're ok.

Comment (amitgalitz) [reply public/expressions/overlay_anomalies.ts:76]: I have tried using the dispatch here in a few several ways and I keep coming back to the fact that we need to wrap the class in a provider in order to use our redux methods. The error I get here is: `Uncaught Error: could not find react-redux context value; please ensure the component is wrapped in a <Provider>`. All our other uses of redux is within the parent component `<Main>` which is wrapped by the redux provider.

Comment (amitgalitz) [reply public/expressions/overlay_anomalies.ts:76]: Do you see other ways we could optimize this to make you more confident in just using the client? I believe the current catching mechanism works as if there is an actual error from the client like a network issue it gets caught in the catch and I have tested this (with mistyping the API for example which gave an error) and if there is some other reason the name can't be returned (I doubt this will actually happen) then that also gets caught.

Comment (ohltyler) [reply public/expressions/overlay_anomalies.ts:76]: Great - I think that's fine to just use the client here. Given that this is a unique context not wrapped in the provider, I think that's ok.

As long as we have sufficient error catching & handling I'm ok with this

**Review by ohltyler — APPROVED** [review]

**Review by jackiehanyang — APPROVED** [review]

**Comment by jackiehanyang** [comment]
some unit tests are failing, other than that, lgtm

**Comment by amitgalitz** [comment]
Unit tests only failing on macos and it passed locally. Not able to re-run tests on this PR.

**Pull request merged**
